src/h3/mod.rs

```diff
@@ -26,4 +26,830 @@
+
+        frame.to_bytes(&mut b).unwrap();
+        let off = b.off();
+        let stream_id = self.get_control_stream_id();
+
+        quic_conn
+            .stream_send(stream_id, &mut d[..off], false)
+            .unwrap();
+    }
+
+    // Send a request
+    pub fn send_request(
+        &mut self, request: http::Request<()>, quic_conn: &mut super::Connection,
+    ) {
+        let mut d = [42; 65535];
+
+        let req_frame = frame::H3Frame::Headers {
+            header_block: H3Util::req_hdrs_to_qpack(
+                &mut self.qpack_encoder,
+                &request,
+            ),
+        };
+
+        let mut b = octets::Octets::with_slice(&mut d);
+        req_frame.to_bytes(&mut b).unwrap();
+        let off = b.off();
+
+        // TODO get an available stream number
+        quic_conn.stream_send(0, &mut d[..off], true).unwrap();
+    }
+
+    // Send a response
+    pub fn send_response(
+        &mut self, stream_id: u64, response: http::Response<()>,
+        quic_conn: &mut super::Connection,
+    ) {
+        let mut d = [42; 65535];
+
+        let headers = frame::H3Frame::Headers {
+            header_block: H3Util::resp_hdrs_to_qpack(
+                &mut self.qpack_encoder,
+                &response,
+            ),
+        };
+
+        let mut b = octets::Octets::with_slice(&mut d);
+        headers.to_bytes(&mut b).unwrap();
+
+        // TODO figure out type management of Response object
+        // theory is that if there is no type, there can be no body
+        // if !response.body().is_empty() {
+        // let data = frame::H3Frame::Data {
+        // payload: response.body().as_bytes().to_vec()
+        // };
+        // data.to_bytes(&mut b).unwrap();
+        // }
+
+        let off = b.off();
+
+        info!(
+            "{} sending response of size {} on stream {}",
+            quic_conn.trace_id(),
+            off,
+            stream_id
+        );
+
+        if let Err(e) = quic_conn.stream_send(stream_id, &mut d[..off], true) {
+            error!("{} stream send failed {:?}", quic_conn.trace_id(), e);
+        }
+    }
+
+    pub fn handle_stream(
+        &mut self, stream_id: u64, quic_conn: &mut super::Connection,
+    ) -> Result<()> {
+        let mut buf = [0; 65535];
+        let (read, fin) = quic_conn.stream_recv(stream_id, &mut buf)?;
+        let mut stream_buf = &mut buf[..read];
+
+        let stream_buf_len = stream_buf.len();
+        info!(
+            "{} stream {} has {} bytes (fin? {})",
+            quic_conn.trace_id(),
+            stream_id,
+            stream_buf_len,
+            fin
+        );
+
+        // H3 unidirectional streams have types as first byte
+        if !stream::is_bidi(stream_id) {
+            let mut o = octets::Octets::with_slice(&mut stream_buf);
+            while o.off() < stream_buf_len {
+                // trace!("loop de loop");
+                // dbg!(&stream_buf);
+                if o.off() == 0 {
```

Comment:
This has the same problem of needing to be at the start of the buffer. But the stream type might be read first, then another stream read is done because data arrives later, and you end up parsing a frame as stream type.